### ISSUE TITLE
Add FTP backend

### ISSUE THREAD

**Issue by pjbull** [drivendataorg/cloudpathlib#26]
Should support both ftp and sftp (ftp over ssh).

There's some code that will help with this in the `zamba-web` project

**Comment by MattOates** [comment]
@pjbull are you interested in a Paramiko based SFTP implementation? I need this right now and would prefer to add it to cloudpathlib than use cloudpathlib plus my own code on the side.

**Comment by pjbull** [comment]
@MattOates yeah, would definitely be interested. It may be helpful to look at the WIP http version here:
https://github.com/drivendataorg/cloudpathlib/pull/468

We likely want the same scenario where to run tests we spin up a server in a separate thread. We also suspect we will need to expose a bit more of the internals like we do for HTTP so people can customize how things are handled for their FTP server, since in my experience it can be idiosyncratic.

Thanks!

**Comment by MattOates** [comment]
@pjbull already got your branch out, thanks. This will be SFTP specifically though, not FTPS so the SSH protocol not FTP over SSL. There should ultimately be almost zero idiosyncrasies other than the auth method being password or keys etc. I have a feeling key based auth might be a bit of a challenge given there is nothing in a URL to specify it. I'll at least implement user-password via URLs "sftp://user:password@host:port/path" to get going.

**Comment by pjbull** [comment]
Great, glad to hear it.

> key based auth might be a bit of a challenge given there is nothing in a URL to specify it.

I think that folks will generally need to do this by either (1) instantiating a client and setting that to the default or using it explicitly (see [examples in the docs](https://cloudpathlib.drivendata.org/stable/authentication/#advanced-use)).  Or, I think we should probably also support setting environment variables for the paths to they key files that get used.
